Base/src/main/java/io/deephaven/base/ringbuffer/AggregatingCharRingBuffer.java

```diff
@@ -0,0 +1,712 @@
+
+            // Evaluate the two unique ranges in the tree.
+            evaluateTwoRanges(offset + (r1h / 2), offset + (r1t / 2),
+                    offset + (r2h / 2), offset + (r2t / 2));
+            return;
+        }
+        if (r1TailNormal <= internalBuffer.storage.length) {
+            // r2 wraps, but r1 does not.
+            final int r1h = 0;
+            final int r1t = r2TailNormal - internalBuffer.storage.length - 1; // change to inclusive
+            final int r2h = r1HeadNormal;
+            final int r2t = r1TailNormal - 1; // change to inclusive
+            final int r3h = r2HeadNormal;
+            final int r3t = internalBuffer.storage.length - 1; // change to inclusive
+
+            // Evaluate the values from the storage buffer and store into the evaluation tree.
+            computeAndStoreResults(r1h, r1t, internalBuffer.storage, offset);
+            computeAndStoreResults(r2h, r2t, internalBuffer.storage, offset);
+            computeAndStoreResults(r3h, r3t, internalBuffer.storage, offset);
+
+            // Evaluate the three unique ranges in the tree.
+            evaluateThreeRanges(offset + (r1h / 2), offset + (r1t / 2),
+                    offset + (r2h / 2), offset + (r2t / 2),
+                    offset + (r3h / 2), offset + (r3t / 2));
+            return;
+        }
+        // r1 wraps, r2 does not.
+        final int r1h = 0;
+        final int r1t = r1TailNormal - internalBuffer.storage.length - 1; // change to inclusive
+        final int r2h = r2HeadNormal;
+        final int r2t = r2TailNormal - 1; // change to inclusive
+        final int r3h = r1HeadNormal;
+        final int r3t = internalBuffer.storage.length - 1; // change to inclusive
+
+        // Evaluate the values from the storage buffer and store into the evaluation tree.
+        computeAndStoreResults(r1h, r1t, internalBuffer.storage, offset);
+        computeAndStoreResults(r2h, r2t, internalBuffer.storage, offset);
+        computeAndStoreResults(r3h, r3t, internalBuffer.storage, offset);
+
+        // Evaluate the three unique ranges in the tree.
+        evaluateThreeRanges(offset + (r1h / 2), offset + (r1t / 2),
+                offset + (r2h / 2), offset + (r2t / 2),
+                offset + (r3h / 2), offset + (r3t / 2));
+    }
+
+    /**
+     * This function accepts three unique ranges in the evaluation tree and iteratively computes aggregation results of
+     * value pairs in the range which are stored in parent nodes. Each iteration will shrink the ranges by a factor of
+     * 2. As the ranges shrink, an overlap will occur and this function will transfer to
+     * {@link #evaluateTwoRanges(int, int, int, int)} for better performance.
+     */
+    private void evaluateThreeRanges(int r1h, int r1t, int r2h, int r2t, int r3h, int r3t) {
+        while (true) {
+            if (r1t >= r2h) {
+                // r1 and r2 overlap, collapse them together and call the two range version.
```

Comment:
period rather than comma, or "so"

```
r1 and r2 overlap. Collapse them together
```

or

```
r1 and r2 overlap, so collapse them together
```